src/coreclr/tools/aot/ILCompiler.ReadyToRun/ObjectWriter/R2RPEBuilder.cs

```diff
@@ -599,12 +599,19 @@ protected override BlobBuilder SerializeSection(string name, SectionLocation loc
 
             if (!_target.IsWindows)
             {
+                const int RVAAlign = 1 << RVABitsToMatchFilePos;
                 if (outputSectionIndex > 0)
                 {
                     sectionStartRva = Math.Max(sectionStartRva, _sectionRVAs[outputSectionIndex - 1] + _sectionRawSizes[outputSectionIndex - 1]);
+
+                    // when assembly is stored in a singlefile bundle, an additional skew is introduced
+                    // as the streams inside the bundle are not necessarily page aligned as we do not 
+                    // know the actual page size on the target system. 
+                    // We may need one page gap of unused VA space before the next section starts.
+                    // We will assume the page size is <= RVAAlign
+                    sectionStartRva += RVAAlign;
```

Comment:
I think I understand the layout mechanism more now. The way we construct sectionStartRva guarantees that sections will not overlap - that is if the file starts form offset 0.
That is not the case in a singlefile bundle. 

On ARM64 the file from a bundle will start at an offset that is 4K aligned. On systems with 4K page size (i.e. default Ubuntu) such aligned offset will be as good as starting form 0. Thus we do not see asserts on Linux.

On OSX ARM64 the page size is 16K, so we may need to "spill" past the end of the target VA region. Same could happen on unusually configured Linuxes with larger page sizes.

I think we can assume that 64K is enough to cover any reasonable page size and leaving 64K unused VA gap between sections will ensure that we will not overlap.